File: sandcastle/sdk-python/src/sandcastle/sandbox.py
from __future__ import annotations

from typing import Any, Optional

import httpx

from sandcastle.errors import SandcastleError
from sandcastle.types import ExecResources, ExecResult


class Sandbox:
    def __init__(self, client: httpx.AsyncClient, base: str, sandbox_id: str, api_key: str):
        self._client = client
        self._base = base.rstrip("/")
        self.id = sandbox_id
        self._api_key = api_key

    def _headers(self) -> dict[str, str]:
        h = {"Content-Type": "application/json"}
        if self._api_key:
            h["Authorization"] = f"Bearer {self._api_key}"
        return h

    async def exec(self, command: str, timeout_secs: Optional[int] = None) -> ExecResult:
        body: dict = {"command": command}
        if timeout_secs is not None:
            body["timeoutSecs"] = timeout_secs
        r = await self._client.post(
            f"{self._base}/v1/sandboxes/{self.id}/exec",
            headers=self._headers(),
            json=body,
        )
        data = r.json() if r.content else {}
        if r.is_error:
            raise SandcastleError(data.get("error", r.text), r.status_code, data)
        res = data.get("resources") or {}
        return ExecResult(
            stdout=data.get("stdout", ""),
            stderr=data.get("stderr", ""),
            exit_code=int(data.get("exitCode", -1)),
            resources=ExecResources(
                memory_peak_mb=int(res.get("memoryPeakMb", 0)),
                cpu_time_ms=int(res.get("cpuTimeMs", 0)),
                wall_time_ms=int(res.get("wallTimeMs", 0)),
                disk_used_mb=int(res.get("diskUsedMb", 0)),
            ),
        )

    async def write_file(self, path: str, content: str | bytes) -> None:
        p = path.lstrip("/")
        r = await self._client.put(
            f"{self._base}/v1/sandboxes/{self.id}/files/{p}",
            headers=self._headers(),
            content=content if isinstance(content, bytes) else content.encode(),
        )
        if r.is_error:
            d = r.json() if r.content else {}
            raise SandcastleError(d.get("error", r.text), r.status_code, d)

    async def read_file(self, path: str) -> str:
        p = path.lstrip("/")
        r = await self._client.get(
            f"{self._base}/v1/sandboxes/{self.id}/files/{p}",
            headers=self._headers(),
        )
        if r.is_error:
            d = r.json() if r.content else {}
            raise SandcastleError(d.get("error", r.text), r.status_code, d)
        return r.text

    async def list_files(self, path: str = "") -> list[dict[str, Any]]:
        q = f"?path={path}" if path else ""
        r = await self._client.get(
            f"{self._base}/v1/sandboxes/{self.id}/files{q}",
            headers=self._headers(),
        )
        data = r.json()
        if r.is_error:
            raise SandcastleError(data.get("error", r.text), r.status_code, data)
        return data

    async def usage(self) -> dict[str, Any]:
        r = await self._client.get(
            f"{self._base}/v1/sandboxes/{self.id}/usage",
            headers=self._headers(),
        )
        data = r.json()
        if r.is_error:
            raise SandcastleError(data.get("error", r.text), r.status_code, data)
        return data

    async def snapshot(self, label: str | None = None) -> dict[str, Any]:
        r = await self._client.post(
            f"{self._base}/v1/sandboxes/{self.id}/snapshot",
            headers=self._headers(),
            json={"label": label} if label else {},
        )
        data = r.json()
        if r.is_error:
            raise SandcastleError(data.get("error", r.text), r.status_code, data)
        return data

    async def replay(self) -> list[dict[str, Any]]:
        r = await self._client.get(
            f"{self._base}/v1/sandboxes/{self.id}/replay",
            headers=self._headers(),
        )
        data = r.json()
        if r.is_error:
            raise SandcastleError(data.get("error", r.text), r.status_code, data)
        return data

    async def destroy(self) -> None:
        r = await self._client.delete(
            f"{self._base}/v1/sandboxes/{self.id}",
            headers=self._headers(),
        )
        if r.is_error:
            d = r.json() if r.content else {}
            raise SandcastleError(d.get("error", r.text), r.status_code, d)

File: sandcastle/sdk-node/src/sandbox.ts
import type { ExecResult, SandcastleConfig, UsageResponse } from "./types.js";
import { SandcastleError } from "./errors.js";

export class Sandbox {
  constructor(
    private readonly cfg: SandcastleConfig,
    public readonly id: string
  ) {}

  private url(path: string): string {
    const base = (this.cfg.baseUrl ?? "http://127.0.0.1:8787").replace(/\/$/, "");
    return `${base}${path}`;
  }

  private headers(): HeadersInit {
    const h: Record<string, string> = {
      "Content-Type": "application/json",
    };
    if (this.cfg.apiKey) h.Authorization = `Bearer ${this.cfg.apiKey}`;
    return h;
  }

  private authHeaders(): HeadersInit {
    const h: Record<string, string> = {};
    if (this.cfg.apiKey) h.Authorization = `Bearer ${this.cfg.apiKey}`;
    return h;
  }

  private async req<T>(
    method: string,
    path: string,
    body?: unknown
  ): Promise<T> {
    const f = this.cfg.fetchImpl ?? fetch;
    const r = await f(this.url(path), {
      method,
      headers: this.headers(),
      body: body === undefined ? undefined : JSON.stringify(body),
    });
    if (r.status === 204) return undefined as T;
    const text = await r.text();
    let json: unknown = null;
    try {
      json = text ? JSON.parse(text) : null;
    } catch {
      json = text;
    }
    if (!r.ok) {
      const msg =
        typeof json === "object" && json && "error" in json
          ? String((json as { error: string }).error)
          : r.statusText;
      throw new SandcastleError(msg, r.status, json);
    }
    return json as T;
  }

  async exec(command: string, timeoutSecs?: number): Promise<ExecResult> {
    return this.req<ExecResult>("POST", `/v1/sandboxes/${this.id}/exec`, {
      command,
      timeoutSecs,
    });
  }

  async execStream(
    command: string,
    timeoutSecs?: number
  ): Promise<WebSocket> {
    const base = (this.cfg.baseUrl ?? "http://127.0.0.1:8787").replace(
      /^http/,
      "ws"
    );
    const wsUrl = `${base}/v1/sandboxes/${this.id}/exec/stream`;
    const ws = new WebSocket(wsUrl);
    await new Promise<void>((resolve, reject) => {
      ws.onopen = () => resolve();
      ws.onerror = () => reject(new SandcastleError("websocket open failed"));
    });
    ws.send(JSON.stringify({ command, timeoutSecs }));
    return ws;
  }

  async writeFile(path: string, content: string | Uint8Array): Promise<void> {
    const base = (this.cfg.baseUrl ?? "http://127.0.0.1:8787").replace(/\/$/, "");
    const bodyInit: BodyInit =
      typeof content === "string" ? content : content;
    const f = this.cfg.fetchImpl ?? fetch;
    const r = await f(`${base}/v1/sandboxes/${this.id}/files/${path.replace(/^\//, "")}`, {
      method: "PUT",
      headers: this.authHeaders(),
      body: bodyInit,
    });
    if (!r.ok) {
      const t = await r.text();
      throw new SandcastleError(t || r.statusText, r.status);
    }
  }

  async readFile(path: string): Promise<string> {
    const base = (this.cfg.baseUrl ?? "http://127.0.0.1:8787").replace(/\/$/, "");
    const f = this.cfg.fetchImpl ?? fetch;
    const r = await f(`${base}/v1/sandboxes/${this.id}/files/${path.replace(/^\//, "")}`, {
      headers: this.authHeaders(),
    });
    if (!r.ok) throw new SandcastleError(await r.text(), r.status);
    return r.text();
  }

  async listFiles(dir = ""): Promise<{ name: string; isDir: boolean; size: number }[]> {
    const q = dir ? `?path=${encodeURIComponent(dir)}` : "";
    return this.req("GET", `/v1/sandboxes/${this.id}/files${q}`);
  }

  async usage(): Promise<UsageResponse> {
    return this.req("GET", `/v1/sandboxes/${this.id}/usage`);
  }

  async snapshot(label?: string): Promise<{ snapshotId: string }> {
    return this.req("POST", `/v1/sandboxes/${this.id}/snapshot`, { label });
  }

  async replay(): Promise<unknown[]> {
    return this.req("GET", `/v1/sandboxes/${this.id}/replay`);
  }

  async destroy(): Promise<void> {
    await this.req("DELETE", `/v1/sandboxes/${this.id}`);
  }
}
